Follow subscriptions no longer work after "Connection to [url] refused".

PGDriverManagerConnectionSupplier.get() catches every SQLException, wraps it inside a RuntimeExpcetion and throws that exception. PGListener (the caller) catches only SQLExceptions in its internal listenerThread. If any other Exception occurs the thread finishes without any more information and the FactCast no longer listens for postgres notifications. 
As a result of that any old or new follow subscription will not receive any new events after its initial catchup.

The corresponding stacktrace looks like the following (line numbers may wary slightly)
```
java.lang.RuntimeException: Cannot acquire Connection from DriverManager: jdbc:postgresql://factcast-test.c2nhv0ptpb7v.eu-central-1.rds.amazonaws.com:5432/TestDB
	at org.factcast.store.pgsql.internal.listen.PGDriverManagerConnectionSupplier.get(PGDriverManagerConnectionSupplier.java:49)
	at org.factcast.store.pgsql.internal.listen.PGDriverManagerConnectionSupplier.get(PGDriverManagerConnectionSupplier.java:21)
	at org.factcast.store.pgsql.internal.listen.PGListener.lambda$listen$0(PGListener.java:64)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.postgresql.util.PSQLException: Connection to factcast-test.c2nhv0ptpb7v.eu-central-1.rds.amazonaws.com:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:265)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
	at org.postgresql.Driver.makeConnection(Driver.java:450)
	at org.postgresql.Driver.connect(Driver.java:252)
	at org.factcast.store.pgsql.internal.listen.PGDriverManagerConnectionSupplier.get(PGDriverManagerConnectionSupplier.java:44)
	... 3 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.postgresql.core.PGStream.<init>(PGStream.java:68)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:144)
	... 8 more
```

--- a/factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/catchup/PGCatchupFactory.java
+++ b/factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/catchup/PGCatchupFactory.java
@@ -11,9 +11,8 @@ import lombok.NonNull;
 
 public interface PGCatchupFactory {
 
-    PGCatchup create(
-
-            @NonNull SubscriptionRequestTO request, @NonNull PGPostQueryMatcher postQueryMatcher,
+    PGCatchup create(@NonNull SubscriptionRequestTO request,
+            @NonNull PGPostQueryMatcher postQueryMatcher,
             @NonNull SubscriptionImpl<Fact> subscription, @NonNull AtomicLong serial);
 
 }

--- a/factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/listen/PGListener.java
+++ b/factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/listen/PGListener.java
@@ -7,7 +7,6 @@ import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
-import java.util.function.Supplier;
 
 import org.factcast.store.pgsql.internal.PGConstants;
 import org.postgresql.PGNotification;
@@ -38,7 +37,7 @@ import lombok.extern.slf4j.Slf4j;
 @Component
 public class PGListener implements InitializingBean, DisposableBean {
 
-    final @NonNull Supplier<PgConnection> ds;
+    final @NonNull PgConnectionSupplier pgConnectionSupplier;
 
     final @NonNull EventBus eventBus;
 
@@ -46,7 +45,7 @@ public class PGListener implements InitializingBean, DisposableBean {
 
     final AtomicBoolean running = new AtomicBoolean(true);
 
-    Thread listenerThread;
+    private Thread listenerThread;
 
     private int blockingWaitTimeInMillis = 1000 * 60;
 
@@ -60,8 +59,8 @@ public class PGListener implements InitializingBean, DisposableBean {
                 // make sure, we did not miss anything while reconnecting
                 postEvent("scheduled-poll");
 
-                try (PgConnection pc = ds.get()) {
-                    try (PreparedStatement ps = pc.prepareStatement(PGConstants.LISTEN_SQL);) {
+                try (PgConnection pc = pgConnectionSupplier.get()) {
+                    try (PreparedStatement ps = pc.prepareStatement(PGConstants.LISTEN_SQL)) {
                         log.trace("Running LISTEN command");
                         ps.execute();
                     }
@@ -102,6 +101,8 @@ public class PGListener implements InitializingBean, DisposableBean {
         }, "PG Instance Listener");
         listenerThread.setDaemon(true);
         listenerThread.start();
+        listenerThread.setUncaughtExceptionHandler((t, e) -> log.error("thread " + t
+                + " encountered an unhandled exception", e));
         try {
             l.await(15, TimeUnit.SECONDS);
         } catch (InterruptedException e) {
@@ -140,12 +141,12 @@ public class PGListener implements InitializingBean, DisposableBean {
     }
 
     @Override
-    public void afterPropertiesSet() throws Exception {
+    public void afterPropertiesSet() {
         listen();
     }
 
     @Override
-    public void destroy() throws Exception {
+    public void destroy() {
         this.running.set(false);
         if (listenerThread != null) {
             listenerThread.interrupt();

--- a/factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/listen/PgConnectionSupplier.java
+++ b/factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/listen/PgConnectionSupplier.java
@@ -3,7 +3,6 @@ package org.factcast.store.pgsql.internal.listen;
 import java.sql.DriverManager;
 import java.sql.SQLException;
 import java.util.Properties;
-import java.util.function.Supplier;
 
 import javax.inject.Inject;
 import javax.sql.DataSource;
@@ -18,15 +17,16 @@ import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
+@Slf4j
 @Component
 @RequiredArgsConstructor
-@Slf4j
-public class PGDriverManagerConnectionSupplier implements Supplier<PgConnection> {
+public class PgConnectionSupplier {
+
     @NonNull
     private final org.apache.tomcat.jdbc.pool.DataSource ds;
 
     @Inject
-    PGDriverManagerConnectionSupplier(@NonNull DataSource dataSource) {
+    PgConnectionSupplier(@NonNull DataSource dataSource) {
 
         if (dataSource instanceof org.apache.tomcat.jdbc.pool.DataSource) {
             this.ds = (org.apache.tomcat.jdbc.pool.DataSource) dataSource;
@@ -37,16 +37,14 @@ public class PGDriverManagerConnectionSupplier implements Supplier<PgConnection>
         }
     }
 
-    @Override
-    public PgConnection get() {
+    public PgConnection get() throws SQLException {
         try {
-
-            return (PgConnection) DriverManager.getDriver(ds.getUrl()).connect(ds.getUrl(),
-                    buildCredentialProperties(ds));
+            return (PgConnection) DriverManager.getDriver(ds.getUrl())
+                    .connect(ds.getUrl(), buildCredentialProperties(ds));
         } catch (SQLException e) {
             final String msg = "Cannot acquire Connection from DriverManager: " + ds.getUrl();
             log.error(msg, e);
-            throw new RuntimeException(msg, e);
+            throw e;
         }
     }
 

--- a/factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/internal/PGSqlListener0Test.java
+++ b/factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/internal/PGSqlListener0Test.java
@@ -7,10 +7,11 @@ import java.sql.Connection;
 import java.sql.PreparedStatement;
 import java.sql.SQLException;
 import java.util.function.Predicate;
-import java.util.function.Supplier;
 
 import org.factcast.store.pgsql.internal.listen.PGListener;
 import org.factcast.store.pgsql.internal.listen.PGListener.FactInsertionEvent;
+import org.factcast.store.pgsql.internal.listen.PgConnectionSupplier;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.ArgumentCaptor;
@@ -28,7 +29,7 @@ import com.google.common.eventbus.AsyncEventBus;
 public class PGSqlListener0Test {
 
     @Mock
-    Supplier<PgConnection> ds;
+    PgConnectionSupplier ds;
 
     @Mock
     AsyncEventBus bus;
@@ -44,7 +45,7 @@ public class PGSqlListener0Test {
     @Captor
     ArgumentCaptor<PGListener> captor;
 
-    @org.junit.Before
+    @Before
     public void setUp() throws SQLException {
         Mockito.when(ds.get()).thenReturn(conn);
         Mockito.when(conn.prepareStatement(anyString())).thenReturn(ps);

--- a/factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/internal/listen/PgConnectionSupplier0Test.java
+++ b/factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/internal/listen/PgConnectionSupplier0Test.java
@@ -10,17 +10,15 @@ import java.util.Properties;
 
 import javax.sql.DataSource;
 
-import junit.framework.Assert;
-
 import org.apache.tomcat.jdbc.pool.PoolConfiguration;
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mock;
 
 @SuppressWarnings("all")
-public class PGDriverManagerConnectionSupplier0Test {
+public class PgConnectionSupplier0Test {
 
-    PGDriverManagerConnectionSupplier uut;
+    PgConnectionSupplier uut;
 
     @Mock
     private org.apache.tomcat.jdbc.pool.DataSource ds;
@@ -29,14 +27,14 @@ public class PGDriverManagerConnectionSupplier0Test {
     public void prepare() {
 
         initMocks(this);
-        uut = new PGDriverManagerConnectionSupplier(ds);
+        uut = new PgConnectionSupplier(ds);
     }
 
     @Test(expected = IllegalStateException.class)
     public void test_wrongDataSourceImplementation() {
 
         DataSource ds = mock(DataSource.class);
-        new PGDriverManagerConnectionSupplier(ds);
+        new PgConnectionSupplier(ds);
 
         failBecauseExceptionWasNotThrown(IllegalStateException.class);
     }
@@ -61,7 +59,7 @@ public class PGDriverManagerConnectionSupplier0Test {
     @Test(expected = NullPointerException.class)
     public void test_constructor() {
 
-        new PGDriverManagerConnectionSupplier(null);
+        new PgConnectionSupplier(null);
 
         failBecauseExceptionWasNotThrown(NullPointerException.class);
 
@@ -73,15 +71,8 @@ public class PGDriverManagerConnectionSupplier0Test {
         org.apache.tomcat.jdbc.pool.DataSource ds = new org.apache.tomcat.jdbc.pool.DataSource();
         ds.setUrl(url);
 
-        PGDriverManagerConnectionSupplier uut = new PGDriverManagerConnectionSupplier(ds);
-        try {
-            uut.get();
-            org.junit.Assert.fail("Was expecting Exception");
-        } catch (Exception e) {
-            if (!(e.getCause() instanceof SQLException)) {
-                fail("Wrong Exception type. Was expecting SQLException wrapped in a RuntimeException, but got ",
-                        e);
-            }
-        }
+        PgConnectionSupplier uut = new PgConnectionSupplier(ds);
+
+        assertThatThrownBy(() -> uut.get()).isInstanceOf(SQLException.class);
     }
 }

